packages/pdfrx/lib/src/widgets/pdf_viewer.dart

```diff
@@ -1023,11 +1533,359 @@ class _PdfViewerState extends State<PdfViewer>
+
+      final viewportPrimarySize = isPrimaryVertical ? _viewSize!.height : _viewSize!.width;
+
+      // Calculate extension distance based on viewport size (not page size)
+      // This ensures proper extension even when viewport is much larger than page (discrete mode with spacing)
+      final extensionDistance = viewportPrimarySize * extensionRatio;
+
+      // Extend to previous page (if exists and should extend)
+      if (shouldExtendToPrev && currentPageNumber > 1) {
+        final prevPageBounds = layout.pageLayouts[currentPageNumber - 2];
+        if (isPrimaryVertical) {
+          // Extend upward
+          final newTop = max(prevPageBounds.top, result.top - extensionDistance);
+          result = Rect.fromLTRB(result.left, newTop, result.right, result.bottom);
+        } else {
+          // Extend leftward
+          final newLeft = max(prevPageBounds.left, result.left - extensionDistance);
+          result = Rect.fromLTRB(newLeft, result.top, result.right, result.bottom);
+        }
+      }
+
+      // Extend to next page (if exists and should extend)
+      if (shouldExtendToNext && currentPageNumber < layout.pageLayouts.length) {
+        final nextPageBounds = layout.pageLayouts[currentPageNumber];
+        if (isPrimaryVertical) {
+          // Extend downward
+          final newBottom = min(nextPageBounds.bottom, result.bottom + extensionDistance);
+          result = Rect.fromLTRB(result.left, result.top, result.right, newBottom);
+        } else {
+          // Extend rightward
+          final newRight = min(nextPageBounds.right, result.right + extensionDistance);
+          result = Rect.fromLTRB(result.left, result.top, newRight, result.bottom);
+        }
+      }
+    }
+    return result;
+  }
+
   // Auto-adjust boundaries when content is smaller than the view, centering
   // the content and ensuring InteractiveViewer's scrollPhysics works when specified
   void _adjustBoundaryMargins(Size viewSize, double zoom) {
     final boundaryMargin = widget.params.boundaryMargin ?? EdgeInsets.zero;
 
+    // Discrete mode: restrict scrolling to current page/spread only
+    if (widget.params.pageTransition == PageTransition.discrete) {
+      final layout = _layout;
+      final currentPageNumber = _pageNumber;
+
+      if (layout == null ||
+          currentPageNumber == null ||
+          currentPageNumber < 1 ||
+          currentPageNumber > layout.pageLayouts.length) {
+        _adjustedBoundaryMargins = boundaryMargin;
+        return;
+      }
+
+      // TODO: should we need _implicit margins if not SpreadLayout?
+      // If we don't need it, I once remove getPageRectWithMargins function.
+      // Rect pageBounds;
+      // if (layout is PdfSpreadLayout) {
+      //   pageBounds = layout.getSpreadBounds(currentPageNumber);
+      // } else {
+      //   pageBounds = layout.getPageRectWithMargins(currentPageNumber);
+      // }
+      final pageBounds = layout.getSpreadBounds(currentPageNumber);
```

Comment:
@enhancient
On your code, if the mode is not spread-layout, you called getPageRectWithMargins. Should we do that??